IPFS: layer not found - follow-up to #3425

### Description

The patch in #3435 did not address the case of IPFS push, which may be a specialized code path, or at least require special treatment compared to normal push.

```
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	84.403s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	87.423s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	84.780s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	76.292s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	78.071s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	74.118s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	93.913s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	74.331s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	77.883s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	77.921s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	75.057s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	81.912s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	75.338s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	72.620s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	84.847s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	72.791s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	87.423s
ok  	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	73.117s
test target: "nerdctl"
time="2024-10-02T18:12:25-07:00" level=error msg="`buildctl` needs to be installed and `buildkitd` needs to be running, see https://github.com/moby/buildkit , and `containerd-rootless-setuptool.sh install-buildkit` for OCI worker or `containerd-rootless-setuptool.sh install-buildkit-containerd` for containerd worker" error="failed to ping to host unix:///run/user/501/buildkit-nerdctl-test/buildkitd.sock: exit status 1\nfailed to ping to host unix:///run/user/501/buildkit-default/buildkitd.sock: exit status 1\nfailed to ping to host unix:///run/user/501/buildkit/buildkitd.sock: exit status 1"
--- FAIL: TestIPFSComposeUpNoBuildDefault (5.49s)
    helpers.go:40: /usr/local/bin/nerdctl --namespace=nerdctl-test pull --quiet ghcr.io/stargz-containers/registry:2-org
    helpers.go:40: /usr/local/bin/nerdctl --namespace=nerdctl-test pull --quiet ghcr.io/stargz-containers/cesanta/docker_auth:1.7-org
    helpers.go:40: /usr/local/bin/nerdctl --namespace=nerdctl-test pull --quiet ghcr.io/stargz-containers/ipfs/kubo:v0.16.0-org
    helpers.go:40: /usr/local/bin/nerdctl --namespace=nerdctl-test rm -f testipfscomposeupnobuilddefault-kubo-registry-server-5000-false-25b9f3c7
    helpers.go:40: /usr/local/bin/nerdctl --namespace=nerdctl-test run --pull=never -d -p 0.0.0.0:5000:5000 --name testipfscomposeupnobuilddefault-kubo-registry-server-5000-false-25b9f3c7 --entrypoint=/bin/sh ghcr.io/stargz-containers/ipfs/kubo:v0.16.0-org -c -- ipfs init && ipfs config Addresses.API /ip4/0.0.0.0/tcp/5000 && ipfs daemon --offline
    cesanta.go:93: /usr/local/bin/nerdctl --namespace=nerdctl-test container inspect testipfscomposeupnobuilddefault-kubo-registry-server-5000-false-25b9f3c7
    helpers.go:60: /usr/local/bin/nerdctl --namespace=nerdctl-test logs testipfscomposeupnobuilddefault-kubo-registry-server-5000-false-25b9f3c7
    helpers.go:40: /usr/local/bin/nerdctl --namespace=nerdctl-test pull --quiet ghcr.io/stargz-containers/wordpress:5.7-org
    helpers.go:40: /usr/local/bin/nerdctl --namespace=nerdctl-test pull --quiet ghcr.io/stargz-containers/mariadb:10.5-org
    ipfs_registry_linux_test.go:39: /usr/local/bin/nerdctl --namespace=nerdctl-test push ipfs://ghcr.io/stargz-containers/wordpress:5.7-org
    ipfs_registry_linux_test.go:39: assertion failed: expect.ExitCode is not result.ExitCode: Expected exit code: 0

        Command:  /usr/local/bin/nerdctl --namespace=nerdctl-test push ipfs://ghcr.io/stargz-containers/wordpress:5.7-org
        ExitCode: 1
        Error:    exit status 1
        Stdout:
        Stderr:   time="2024-10-02T18:12:29-07:00" level=info msg="pushing image \"ghcr.io/stargz-containers/wordpress:5.7-org\" to IPFS"
        time="2024-10-02T18:12:29-07:00" level=info msg="ensuring image contents"
        time="2024-10-02T18:12:30-07:00" level=warning msg="ipfs push failed" error="content digest sha256:5625a2e212348c7f566e6686e3706a49f9ba12eb8f129c74fcc5b730939de178: not found"
        time="2024-10-02T18:12:30-07:00" level=fatal msg="content digest sha256:5625a2e212348c7f566e6686e3706a49f9ba12eb8f129c74fcc5b730939de178: not found"

        Env:
        SHELL=/bin/bash
        LOGNAME=dmp
        XDG_SESSION_TYPE=tty
        HOME=/home/dmp.linux
        LANG=C.UTF-8
        SSH_CONNECTION=192.168.5.2 51297 192.168.5.15 22
        LESSCLOSE=/usr/bin/lesspipe %s %s
        XDG_SESSION_CLASS=user
        TERM=xterm-256color
        LESSOPEN=| /usr/bin/lesspipe %s
        USER=dmp
        SHLVL=2
        XDG_SESSION_ID=2
        XDG_RUNTIME_DIR=/run/user/501
        SSH_CLIENT=192.168.5.2 51297 22
        XDG_DATA_DIRS=/usr/local/share:/usr/share:/var/lib/snapd/desktop
        DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/501/bus
        SSH_TTY=/dev/pts/1
        OLDPWD=/Users/dmp
        _=/usr/local/go/bin/go
        PATH=/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/sbin:/sbin:/usr/sbin:/sbin:/usr/sbin:/sbin:/usr/local/go/bin
        PWD=/Users/dmp/Projects/go/nerd/nerdctl/cmd/nerdctl/ipfs
        IPFS_PATH=/home/dmp.linux/.local/share/ipfs
        DOCKER_CONFIG=/tmp/TestIPFSComposeUpNoBuildDefault1952484318/001
        NERDCTL_TOML=/tmp/TestIPFSComposeUpNoBuildDefault1952484318/001/nerdctl.toml
    helpers.go:47: /usr/local/bin/nerdctl --namespace=nerdctl-test rm -f testipfscomposeupnobuilddefault-kubo-registry-server-5000-false-25b9f3c7
    helpers.go:47: /usr/local/bin/nerdctl --namespace=nerdctl-test rmi
    helpers.go:47: /usr/local/bin/nerdctl --namespace=nerdctl-test rmi
time="2024-10-02T18:12:31-07:00" level=error msg="`buildctl` needs to be installed and `buildkitd` needs to be running, see https://github.com/moby/buildkit , and `containerd-rootless-setuptool.sh install-buildkit` for OCI worker or `containerd-rootless-setuptool.sh install-buildkit-containerd` for containerd worker" error="failed to ping to host unix:///run/user/501/buildkit-nerdctl-test/buildkitd.sock: exit status 1\nfailed to ping to host unix:///run/user/501/buildkit-default/buildkitd.sock: exit status 1\nfailed to ping to host unix:///run/user/501/buildkit/buildkitd.sock: exit status 1"
FAIL
FAIL	github.com/containerd/nerdctl/v2/cmd/nerdctl/ipfs	80.191s
FAIL
FAIL
```

### Steps to reproduce the issue

Hammer the tests repeatedly until they bleed...

### Describe the results you received and expected

na

### What version of nerdctl are you using?

main

### Are you using a variant of nerdctl? (e.g., Rancher Desktop)

None

### Host information

_No response_

--- a/cmd/nerdctl/issues/issues_linux_test.go
+++ b/cmd/nerdctl/issues/issues_linux_test.go
@@ -20,8 +20,13 @@ package issues
 
 import (
 	"fmt"
+	"os"
+	"path/filepath"
 	"testing"
 
+	"gotest.tools/v3/assert"
+
+	"github.com/containerd/nerdctl/v2/pkg/rootlessutil"
 	"github.com/containerd/nerdctl/v2/pkg/testutil"
 	"github.com/containerd/nerdctl/v2/pkg/testutil/nerdtest"
 	"github.com/containerd/nerdctl/v2/pkg/testutil/test"
@@ -36,6 +41,16 @@ func TestIssue3425(t *testing.T) {
 	nerdtest.Setup()
 
 	var registry *testregistry.RegistryServer
+
+	var ipfsPath string
+	if rootlessutil.IsRootless() {
+		var err error
+		ipfsPath, err = rootlessutil.XDGDataHome()
+		ipfsPath = filepath.Join(ipfsPath, "ipfs")
+		assert.NilError(t, err)
+	} else {
+		ipfsPath = filepath.Join(os.Getenv("HOME"), ".ipfs")
+	}
 
 	testCase := &test.Case{
 		Description: "TestIssue3425",
@@ -123,6 +138,32 @@ func TestIssue3425(t *testing.T) {
 				},
 				Command: func(data test.Data, helpers test.Helpers) test.Command {
 					return helpers.Command("image", "convert", "--oci", "--estargz", testutil.CommonImage, data.Identifier())
+				},
+				Expected: test.Expects(0, nil, nil),
+			},
+			{
+				Description: "with ipfs",
+				Require: test.Require(
+					nerdtest.Private,
+					test.Not(test.Windows),
+					test.Not(nerdtest.Docker),
+					test.Binary("ipfs"),
+				),
+				Env: map[string]string{
+					"IPFS_PATH": ipfsPath,
+				},
+				Setup: func(data test.Data, helpers test.Helpers) {
+					helpers.Ensure("image", "pull", testutil.CommonImage)
+					helpers.Ensure("run", "-d", "--name", data.Identifier(), testutil.CommonImage)
+					helpers.Ensure("image", "rm", "-f", testutil.CommonImage)
+					helpers.Ensure("image", "pull", testutil.CommonImage)
+				},
+				Cleanup: func(data test.Data, helpers test.Helpers) {
+					helpers.Anyhow("rm", "-f", data.Identifier())
+					helpers.Anyhow("rmi", "-f", data.Identifier())
+				},
+				Command: func(data test.Data, helpers test.Helpers) test.Command {
+					return helpers.Command("image", "push", "ipfs://"+testutil.CommonImage)
 				},
 				Expected: test.Expects(0, nil, nil),
 			},

--- a/pkg/cmd/image/ensure.go
+++ b/pkg/cmd/image/ensure.go
@@ -19,7 +19,6 @@ package image
 import (
 	"context"
 	"errors"
-	"fmt"
 	"net/http"
 	"os"
 
@@ -43,7 +42,6 @@ func EnsureAllContent(ctx context.Context, client *containerd.Client, srcName st
 	imageService := client.ImageService()
 	img, err := imageService.Get(ctx, srcName)
 	if err != nil {
-		fmt.Println("Failed getting imageservice")
 		return err
 	}
 

--- a/pkg/cmd/image/push.go
+++ b/pkg/cmd/image/push.go
@@ -61,6 +61,17 @@ func Push(ctx context.Context, client *containerd.Client, rawRef string, options
 		}
 		log.G(ctx).Infof("pushing image %q to IPFS", ref)
 
+		parsedRef, err := distributionref.ParseDockerRef(ref)
+		if err != nil {
+			return err
+		}
+
+		// Ensure all the layers are here: https://github.com/containerd/nerdctl/issues/3489
+		err = EnsureAllContent(ctx, client, parsedRef.String(), options.GOptions)
+		if err != nil {
+			return err
+		}
+
 		var ipfsPath string
 		if options.IpfsAddress != "" {
 			dir, err := os.MkdirTemp("", "apidirtmp")
@@ -78,7 +89,7 @@ func Push(ctx context.Context, client *containerd.Client, rawRef string, options
 		if options.Estargz {
 			layerConvert = eStargzConvertFunc()
 		}
-		c, err := ipfs.Push(ctx, client, ref, layerConvert, options.AllPlatforms, options.Platforms, options.IpfsEnsureImage, ipfsPath)
+		c, err := ipfs.Push(ctx, client, parsedRef.String(), layerConvert, options.AllPlatforms, options.Platforms, options.IpfsEnsureImage, ipfsPath)
 		if err != nil {
 			log.G(ctx).WithError(err).Warnf("ipfs push failed")
 			return err

--- a/pkg/ipfs/image.go
+++ b/pkg/ipfs/image.go
@@ -37,7 +37,6 @@ import (
 	"github.com/containerd/nerdctl/v2/pkg/idutil/imagewalker"
 	"github.com/containerd/nerdctl/v2/pkg/imgutil"
 	"github.com/containerd/nerdctl/v2/pkg/platformutil"
-	"github.com/containerd/nerdctl/v2/pkg/referenceutil"
 )
 
 const ipfsPathEnv = "IPFS_PATH"
@@ -93,11 +92,7 @@ func Push(ctx context.Context, client *containerd.Client, rawRef string, layerCo
 			log.G(ctx).WithError(err).Warnf("failed to ensure the existence of image %q", rawRef)
 		}
 	}
-	ref, err := referenceutil.ParseAny(rawRef)
-	if err != nil {
-		return "", err
-	}
-	return ipfs.PushWithIPFSPath(ctx, client, ref.String(), layerConvert, platMC, &ipath)
+	return ipfs.PushWithIPFSPath(ctx, client, rawRef, layerConvert, platMC, &ipath)
 }
 
 // ensureContentsOfIPFSImage ensures that the entire contents of an existing IPFS image are fully downloaded to containerd.

